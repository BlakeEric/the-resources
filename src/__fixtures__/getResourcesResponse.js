const getResourcesResponse = [
  {
    node: {
      id: 'd04da249-50e7-5b18-a992-d530e14831a7',
      frontmatter: {
        title: 'A List Apart',
        format: 'Article',
        focus: 'Front-end',
        skillLevel: ['Beginner','Moderate','Advanced']
      },
      fields: {
        slug: '/a-list-apart/'
      }
    }
  },
  {
    node: {
      id: '77a6be63-d3c7-50cb-bd29-438427560388',
      frontmatter: {
        title: 'Scrimba',
        format: 'Interactive',
        focus: 'Front-end',
        skillLevel: ['Beginner','Moderate']
      },
      fields: {
        slug: '/scrimba/'
      }
    }
  },
  {
    node: {
      id: '28feb8c5-acb2-5638-bdd6-b0457eff8db3',
      frontmatter: {
        title: 'Smashing Magazine',
        format: 'Article',
        focus: 'Multiple',
        skillLevel: ['Beginner','Moderate','Advanced']
      },
      fields: {
        slug: '/smashing-magazine/'
      }
    }
  },
  {
    node: {
      id: '744d8211-8c35-5c83-b8c1-ecb4ab48c0db',
      frontmatter: {
        title: 'The Odin Project',
        format: 'Tutorial',
        focus: 'Multiple',
        skillLevel: ['Beginner','Moderate']
      },
      fields: {
        slug: '/the-odin-project/'
      }
    }
  },
  {
    node: {
      id: '3b19247d-2ddd-558d-b0ce-14ec1972d52f',
      frontmatter: {
        title: 'Watch and Code',
        format: 'Video',
        focus: 'Front-end',
        skillLevel: ['Beginner']
      },
      fields: {
        slug: '/watch-and-code/'
      }
    }
  }
]

export default getResourcesResponse;
