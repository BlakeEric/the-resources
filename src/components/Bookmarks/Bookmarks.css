.bookmarks {
	border-top: 1px solid lightgray;    
	margin-top: 1rem;
}

.bookmarks h3 {
	margin-bottom: 0;
	margin-top: 1rem;
}

.bookmarks p {
	color: gray;
	margin: 0;
}

.bookmarks ul {
	list-style: none;
	padding-left: 0;
	margin-top: 0;
}

.bookmarks ul li a {
	padding-right: 21px;
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
	display: inline-block;
	max-width: 100%;
}

.bookmarks .btn-remove {
	color: darkred;
	background: none;
	padding: 0;
	line-height: 1;
	font-size: 125%;
	position: absolute;
	right: 5px;
	height: 10px;
	top: 3px;
	box-shadow: none;
} 

.bookmarks .btn-remove:hover, 
.bookmarks .btn-remove:focus {
	background: none;
}