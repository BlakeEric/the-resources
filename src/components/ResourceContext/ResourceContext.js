import React from 'react'


export const ResourceContext = React.createContext();


export class ResourceProvider extends React.Component {

  state = {
    bookmarks: [],
    currentToggledItemId: null,
    isFilterable: false, // used to display filter options if JS is enabled
    orderBy: "name",
    order: "ASC",
    resources: this.props.resources,
    searchTerm: null,
    selectedSkillLevel: null,
    selectedFormat: null,
    selectedFocus: null,
  }


  componentDidMount() {
    // if JS is enabled allow dynamic filtering
    this.setState({isFilterable: true})
  }


  /*
   * add an resource's id to bookmarks of saved resources
   */
  addToBookmarks = (id) => {
    this.setState({
      bookmarks: [...this.state.bookmarks, id]
    })
  }


  /*
   * remove an resource's id from bookmarks of saved resources
   */
  removeFromBookmarks = (id) => {
    this.setState({
      bookmarks: this.state.bookmarks.filter(item => item !== id)
    })
  }


  /*
   * Reverse value of 'order' in state: "ASC" or "DESC"
   */
  reverseOrder = () => {
    let currentOrder = this.state.order;
    let order;

    if (currentOrder === "DESC") {
      order = "ASC"
    } else {
      order = "DESC"
    }

    this.setState({
      order
    })
  }

  setToggledItemId = (id) => {
    this.setState({
      currentToggledItemId: id
    })
  }

  /*
   * Set the key to use as basis for alphabetical sorting
   */
  setOrderBy = (key) => {
    this.setState({
      orderBy: key,
      order: "ASC"
    })
  }


  /*
   * Click handler that reverses order in state if a new field is clicked
   * Otherwise sets orderBy value to new key
   */
  setOrder = (key) => {
    if (!this.state.isFilterable) {
      return;
    }

    if (this.state.orderBy === key) {
      this.reverseOrder();
    } else {
      this.setOrderBy(key);
    }
  }


  /*
   * Set search term in state
   */
  setSearchTerm = (searchTerm) => {
    this.setState({searchTerm})
  }


  /*
   * Set filters in state
   */
  setSelected = (key, value) => {
    this.setState({[key]: value})
  }


  /*
   * Clear all selected filters from state
   */
  clearFilters = () => {
    this.setState({
      searchTerm: null,
      selectedSkillLevel: null,
      selectedFormat: null,
      selectedFocus: null
    })
  }


  /*
   * Alphabetically compare one field against another
   * Return -1, 0, or 1 for use in Array.sort
   */
  compareFields(current, next) {

    if (this.state.order === "DESC") {

      if(current > next) { return -1; }
      else if(current < next) { return 1; }

    } else {

      if(current < next) { return -1; }
      else if(current > next) { return 1; }

    }

    return 0;
  }


  /*
   * Filter the resources based on orderBy and search values set in state
   * Return reordered array
   */
  filter = () => {

    const { searchTerm, selectedSkillLevel, selectedFocus, selectedFormat } = this.state;
    const key = this.state.orderBy || 'name';

    // create a new array of resources to avoid mutating the value in state
    let filteredItems = [...this.state.resources];

    // remove resources not matching the search term in state
    if (searchTerm) {
      let string = searchTerm.toLowerCase()

      filteredItems = filteredItems.filter(item => {
        return item.node.frontmatter.name.toLowerCase().includes(string)
      })
    }

    // remove resources not matching the skillLevel in state
    if (selectedSkillLevel) {
      let string = selectedSkillLevel.toLowerCase()

      filteredItems = filteredItems.filter(item => {
        return item.node.frontmatter.skillLevel.toString().toLowerCase().includes(string)
      })
    }

    // remove resources not matching the format in state
    if (selectedFormat) {
      let string = selectedFormat.toLowerCase()

      filteredItems = filteredItems.filter(item => {
        return item.node.frontmatter.format.toLowerCase().includes(string)
      })
    }

    // remove resources not matching the format in state
    if (selectedFocus) {
      let string = selectedFocus.toLowerCase()

      filteredItems = filteredItems.filter(item => {
        return item.node.frontmatter.focus.toLowerCase().includes(string)
      })
    }

    // Sort the items alphabetically by "orderBy" key set in state.
    // Will sort in ASC order by default, or DESC if set as "order" value in state
    filteredItems.sort((current, next) => {

      let currentField = current.node.frontmatter[key].toString().toLowerCase();
      let nextField = next.node.frontmatter[key].toString().toLowerCase();

      if (key === "skillLevel") {
        currentField = currentField.replace("beginner,moderate,advanced", "all");
        nextField = nextField.replace("beginner,moderate,advanced", "all");
      }

      return this.compareFields(currentField, nextField)

    })

    return filteredItems
  }


  render() {

    const contextValue = {
      ...this.state,
      reverseOrder: this.reverseOrder,
      setToggledItemId: this.setToggledItemId,
      setOrder: this.setOrder,
      setOrderBy: this.setOrderBy,
      getFilteredResources: this.filter,
      setSearchTerm: this.setSearchTerm,
      setSelected: this.setSelected,
      clearFilters: this.clearFilters,
      allFormats: this.props.allFormats,
      allFocuses: this.props.allFocuses,
      allSkillLevels: this.props.allSkillLevels,
      addToBookmarks: this.addToBookmarks,
      removeFromBookmarks: this.removeFromBookmarks
    }

    return (
      <ResourceContext.Provider value={contextValue}>
        {this.props.children}
      </ResourceContext.Provider>
    );
  }
}
