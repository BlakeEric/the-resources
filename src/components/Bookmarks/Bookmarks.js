import React, { useContext } from "react"
import { css } from '@emotion/core'
import { rhythm } from "../../utils/typography"
import { ResourceContext } from '../ResourceContext/ResourceContext'

export default function Bookmarks() {

  const context = useContext(ResourceContext)

  const items = context.resources.filter(item => context.bookmarks.includes(item.node.id))

  return (
    <section className="bookmarks" css={styles}>
      <h3>My Bookmarks</h3>
      {items.length > 0 ?
        <ul>
          {items.map((item, i) =>
            <li key={`bookmarksItem-${i}`}>
              <a className="btn-view" href={item.node.frontmatter.url} target="_blank" rel="noreferrer noopener">{item.node.frontmatter.name}</a> <button className="btn-remove" onClick={() => context.removeFromBookmarks(item.node.id)}>Ã—</button>
            </li>
          )}
        </ul>
      :
        <p>Select and item to bookmark it.</p>
      }

    </section>
  )
}

const styles = css`
  position: sticky;
  top: calc(50% - 20%);
  width: 300px;
  max-width: 100%;
  border-radius: 5px;
  padding: ${rhythm(.25)} ${rhythm(.325)};
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.05);
  border: 1px solid #eaeaea;
  h3 {
    letter-spacing: 1px;
    line-height: 1.5;
    border-radius: 5px;
    text-transform: uppercase;
    font-weight: 300;
    margin-bottom: ${rhythm(.25)};
    font-size: ${rhythm(.5)};
  }
  p {
    margin-bottom: 0;
    font-size: ${rhythm(.525)};
  }
  ul {
    font-size: ${rhythm(.525)};
    list-style: none;
    margin: 0;
    display: block;
  }
  li {
    width: 100%;
    margin: 0;
    display: block;
    border-bottom: 1px solid #eaeaea;
    position: relative;
    line-height: 1.25;
    padding: ${rhythm(.125)} 23px ${rhythm(.25)} 0;
    &:last-child {
      border-bottom: none;
    }
  }
  .btn-remove {
    position: absolute;
    right: 0;
    height: 22px;
    width: 22px;
    background: white;
    border: none;
    padding: 0;
    text-align: center;
    margin-left: ${rhythm(.125)};
    color: #ce3030;
    border-radius: 50%;
    &:hover, &:focus {
      background-image: linear-gradient(to bottom right,#ce3030,#ef6666);
      color: white;
      box-shadow: 0 2px 2px 0 rgba(0,0,0,0.05);
    }
  }
`
