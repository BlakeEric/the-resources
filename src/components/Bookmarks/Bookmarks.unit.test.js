import React from "react"
import renderer from "react-test-renderer"
import Bookmarks from "./Bookmarks"
import getResourcesResponse from '../../__fixtures__/getResourcesResponse'
import { ResourceContext } from "../ResourceContext/ResourceContext"

let mockContext = {
  bookmarks: [],
  resources: getResourcesResponse
}

describe("Bookmarks", () => {
  it("renders correctly when no items bookmarked", () => {
    const tree = renderer.create(
      <ResourceContext.Provider value={mockContext}>
        <Bookmarks />
      </ResourceContext.Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders correctly when context has bookmarks", () => {

    mockContext.bookmarks = [getResourcesResponse[0].node.id, getResourcesResponse[1].node.id]
    const tree = renderer.create(
      <ResourceContext.Provider value={mockContext}>
        <Bookmarks />
      </ResourceContext.Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
