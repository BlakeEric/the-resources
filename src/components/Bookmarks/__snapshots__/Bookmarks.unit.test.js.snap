// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmarks renders correctly when context has bookmarks 1`] = `
<section
  className="bookmarks"
  css={
    Object {
      "map": undefined,
      "name": "1q4azch",
      "next": undefined,
      "styles": "
  position: sticky;
  top: calc(50% - 20%);
  width: 300px;
  max-width: 100%;
  border-radius: 5px;
  padding: 0.4025rem 0.52325rem;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.05);
  border: 1px solid #eaeaea;
  h3 {
    letter-spacing: 1px;
    line-height: 1.5;
    border-radius: 5px;
    text-transform: uppercase;
    font-weight: 300;
    margin-bottom: 0.4025rem;
    font-size: 0.805rem;
  }
  p {
    margin-bottom: 0;
    font-size: 0.84525rem;
  }
  ul {
    font-size: 0.84525rem;
    list-style: none;
    margin: 0;
    display: block;
  }
  li {
    width: 100%;
    margin: 0;
    display: block;
    border-bottom: 1px solid #eaeaea;
    position: relative;
    line-height: 1.25;
    padding: 0.20125rem 23px 0.4025rem 0;
    &:last-child {
      border-bottom: none;
    }
  }
  .btn-remove {
    position: absolute;
    right: 0;
    height: 22px;
    width: 22px;
    background: white;
    border: none;
    padding: 0;
    text-align: center;
    margin-left: 0.20125rem;
    color: #ce3030;
    border-radius: 50%;
    &:hover, &:focus {
      background-image: linear-gradient(to bottom right,#ce3030,#ef6666);
      color: white;
      box-shadow: 0 2px 2px 0 rgba(0,0,0,0.05);
    }
  }
",
      "toString": [Function],
    }
  }
>
  <h3>
    My Bookmarks
  </h3>
  <ul>
    <li>
      <a
        className="btn-view"
        rel="noreferrer noopener"
        target="_blank"
      >
        Alligator
      </a>
       
      <button
        className="btn-remove"
        onClick={[Function]}
      >
        ×
      </button>
    </li>
    <li>
      <a
        className="btn-view"
        rel="noreferrer noopener"
        target="_blank"
      >
        Badger
      </a>
       
      <button
        className="btn-remove"
        onClick={[Function]}
      >
        ×
      </button>
    </li>
  </ul>
</section>
`;

exports[`Bookmarks renders correctly when no items bookmarked 1`] = `
<section
  className="bookmarks"
  css={
    Object {
      "map": undefined,
      "name": "1q4azch",
      "next": undefined,
      "styles": "
  position: sticky;
  top: calc(50% - 20%);
  width: 300px;
  max-width: 100%;
  border-radius: 5px;
  padding: 0.4025rem 0.52325rem;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.05);
  border: 1px solid #eaeaea;
  h3 {
    letter-spacing: 1px;
    line-height: 1.5;
    border-radius: 5px;
    text-transform: uppercase;
    font-weight: 300;
    margin-bottom: 0.4025rem;
    font-size: 0.805rem;
  }
  p {
    margin-bottom: 0;
    font-size: 0.84525rem;
  }
  ul {
    font-size: 0.84525rem;
    list-style: none;
    margin: 0;
    display: block;
  }
  li {
    width: 100%;
    margin: 0;
    display: block;
    border-bottom: 1px solid #eaeaea;
    position: relative;
    line-height: 1.25;
    padding: 0.20125rem 23px 0.4025rem 0;
    &:last-child {
      border-bottom: none;
    }
  }
  .btn-remove {
    position: absolute;
    right: 0;
    height: 22px;
    width: 22px;
    background: white;
    border: none;
    padding: 0;
    text-align: center;
    margin-left: 0.20125rem;
    color: #ce3030;
    border-radius: 50%;
    &:hover, &:focus {
      background-image: linear-gradient(to bottom right,#ce3030,#ef6666);
      color: white;
      box-shadow: 0 2px 2px 0 rgba(0,0,0,0.05);
    }
  }
",
      "toString": [Function],
    }
  }
>
  <h3>
    My Bookmarks
  </h3>
  <p>
    Select and item to bookmark it.
  </p>
</section>
`;
